### PR TITLE
SI-4365 nondeterministic failure in asSeenFrom

### PR DESCRIPTION
**Pull request by paulp**
Under some order-dependent conditions (if source files arrive
in one order it happens, in the other order it does not) more
than one set of type parameters are created for a given class.
Previously this would lead to a crash in asSeenFrom when a type
parameter had to be matched up with a type application.

Now when that situation arises I compare them by name and log
a dev warning if it hits. This does not risk anything undesirable
happening because the wayward type parameter's owner is always
the right class; it's only the class type parameters which don't
include the wayward one. Since in a given type parameter list
names are unique, we have enough information to salvage the
search.

### CODE REVIEW

**Comment by paulp** [comment]
Review by @adriaanm 

don't forget the standing ovation

**Comment by paulp** [comment]
The flurry of little pushes was just me beautifying the error message.

File: src/reflect/scala/reflect/internal/tpe/TypeMaps.scala [hunk 1]
```diff
@@ -525,15 +525,39 @@ private[internal] trait TypeMaps {
...
-        abort(s"something is wrong: cannot make sense of type application\n  $lhs\n  $rhs")
+      else {
+        // It's easy to get here when working on hardcore type machinery (not to
+        // mention when not doing so, see above) so let's provide a standout error.
+        def own_s(s: Symbol) = s.nameString + " in " + s.safeOwner.nameString
+        def explain =
+          s"""|   sought  ${own_s(lhsSym)}
+              | classSym  ${own_s(rhsSym)}
+              |  tparams  ${rhsSym.typeParams map own_s mkString ", "}
+              |""".stripMargin
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/tpe/TypeMaps.scala]: Could use our `sm` interpolator here.

File: src/reflect/scala/reflect/internal/tpe/TypeMaps.scala [hunk 2]
```diff
@@ -525,15 +525,39 @@ private[internal] trait TypeMaps {
...
+
+        if (argIndex < 0)
+          abort("Something is wrong: cannot find $lhs in applied type $rhs\n" + explain)
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/tpe/TypeMaps.scala]: Apparently missing the `s` interpolator.

**Comment by vigdorchik** [comment]
@paulp I see this kind of failure in my interactive scaladoc work in the IDE, that does reloads of the source files of allOverriddenSymbols.  I guess that might be related.

java.lang.Error
        at scala.reflect.internal.Types$AsSeenFromMap.instParam$1(Types.scala:4\
588)
        at scala.reflect.internal.Types$AsSeenFromMap.toInstance$1(Types.scala:\
4611)
        at scala.reflect.internal.Types$AsSeenFromMap.apply(Types.scala:4629)
        at scala.reflect.internal.Types$AsSeenFromMap.apply(Types.scala:4493)
....

Would it be possible to backport to 2.10.x, given that the IDE is going to support it for some time to come?

**Comment by ihji** [comment]
I'm really glad to see this fix :smile:. I couldn't use Intellij for compiling the Scala library source since few weeks ago because of this problem.

**Comment by paulp** [comment]
@vigdorchik I imagine I can backport it, but I'd like to get it into master first.

**Comment by paulp** [comment]
@adriaanm Don't let all that applauding make you forget your LGTMs

**Comment by adriaanm** [comment]
LGTM :clap: :clap: :clap:

**Comment by adriaanm** [comment]
Since this a bit of a black swan of scalac bugs, I'd consider promoting the `devWarning` to a full warning.

**Comment by adriaanm** [comment]
Merging anyway to appease the hungry masses.

**Pull request merged**